chain/network/src/config.rs

```diff
@@ -104,54 +107,75 @@ pub struct NetworkConfig {
 }
 
 impl NetworkConfig {
+    // Constructs and validates the config.
+    // TODO(gprusak): make the output immutable, to enforce the invariants
+    // checked during validation: either make it return an Arc, or add an Inner type,
+    // so that NetworkConfig dereferences to Inner.
     pub fn new(
         cfg: crate::config_json::Config,
         node_key: SecretKey,
         validator_signer: Option<Arc<dyn ValidatorSigner>>,
         archive: bool,
-    ) -> Self {
-        Self {
+    ) -> anyhow::Result<Self> {
+        let this = Self {
             node_key,
-            validator: validator_signer.as_ref().map(|signer| ValidatorConfig {
-                signer: signer.clone(),
-                endpoints: if cfg.public_addrs.len() > 0 {
-                    ValidatorEndpoints::PublicAddrs(
-                        cfg.public_addrs
-                            .into_iter()
-                            .map(|addr| addr.parse().expect("Failed to parse SocketAddr"))
-                            .collect(),
-                    )
-                } else {
-                    ValidatorEndpoints::TrustedStunServers(cfg.trusted_stun_servers)
-                },
-            }),
+            validator: validator_signer
+                .as_ref()
+                .map(|signer| {
+                    anyhow::Ok(ValidatorConfig {
+                        signer: signer.clone(),
+                        endpoints: if cfg.public_addrs.len() > 0 {
+                            let peers: Vec<_> = cfg
+                                .public_addrs
+                                .iter()
+                                .cloned()
+                                .map(|addr| PeerAddr { addr, peer_id: None })
+                                .collect();
+                            // Check that the amount of data is not too large.
+                            AccountData {
+                                peers: peers.clone(),
+                                account_id: signer.validator_id().clone(),
+                                epoch_id: EpochId::default(),
+                                timestamp: time::Clock::real().now_utc(),
+                            }
+                            .sign(signer.deref())?;
+                            ValidatorEndpoints::PublicPeerAddrs(peers)
+                        } else {
+                            ValidatorEndpoints::TrustedStunServers(cfg.trusted_stun_servers)
+                        },
+                    })
+                })
+                .transpose()?,
```

Comment:
This is a particularly busy bit of code, extracting some variables could potentially help.

Reply:
ack, as discussed offline, please propose a refactor in a followup PR if you think it is worth it.